pkg/controller/volume/persistentvolume/scheduler_binder.go

```diff
@@ -253,58 +254,190 @@ func (b *volumeBinder) AssumePodVolumes(assumedPod *v1.Pod, nodeName string) (al
 			return err
 		}
+		unprocessedBindings++
 	}
 
 	// Update claims objects to trigger volume provisioning. Let the PV controller take care of the rest
 	// PV controller is expect to signal back by removing related annotations if actual provisioning fails
-	for i, claim := range claimsToProvision {
+	for _, claim = range claimsToProvision {
+		glog.V(5).Infof("bindAPIUpdate: Pod %q, PVC %q", podName, getPVCName(claim))
 		if _, err := b.ctrl.kubeClient.CoreV1().PersistentVolumeClaims(claim.Namespace).Update(claim); err != nil {
-			glog.V(4).Infof("updating PersistentVolumeClaim[%s] failed: %v", getPVCName(claim), err)
-			// only revert assumed cached updates for claims we haven't successfully updated
-			b.revertAssumedPVCs(claimsToProvision[i:])
 			return err
 		}
+		unprocessedProvisionings++
 	}
 
 	return nil
 }
 
-func getPodName(pod *v1.Pod) string {
-	return pod.Namespace + "/" + pod.Name
+// checkBindings runs through all the PVCs in the Pod and checks:
+// * if the PVC is fully bound
+// * if there are any conditions that require binding to fail and be retried
+//
+// It returns true when all of the Pod's PVCs are fully bound, and error if
+// binding (and scheduling) needs to be retried
+func (b *volumeBinder) checkBindings(podName string, bindings []*bindingInfo, claimsToProvision []*v1.PersistentVolumeClaim) (bool, error) {
+	for _, binding := range bindings {
+		// Check for any conditions that might require scheduling retry
+
+		// Check if pv still exists
+		pv, err := b.pvCache.GetPV(binding.pv.Name)
+		if err != nil || pv == nil {
+			return false, fmt.Errorf("failed to check pv binding: %v", err)
+		}
+
+		// Check if pv.ClaimRef got dropped by unbindVolume()
+		if pv.Spec.ClaimRef == nil || pv.Spec.ClaimRef.UID == "" {
+			return false, fmt.Errorf("ClaimRef got reset for pv %q", pv.Name)
+		}
+
+		// Check if pvc is fully bound
+		if isBound, _, err := b.isPVCBound(binding.pvc.Namespace, binding.pvc.Name, true); !isBound || err != nil {
+			return false, err
+		}
+
+		// TODO; what if pvc is bound to the wrong pv? It means our assume cache should be reverted.
+		// Or will pv controller cleanup the pv.ClaimRef?
+	}
+
+	for _, claim := range claimsToProvision {
+		bound, pvc, err := b.isPVCBound(claim.Namespace, claim.Name, true)
+		if err != nil || pvc == nil {
+			return false, fmt.Errorf("failed to check pvc binding: %v", err)
+		}
+
+		// Check if selectedNode annotation is still set
+		if pvc.Annotations == nil || len(pvc.Annotations[annSelectedNode]) == 0 {
+			return false, fmt.Errorf("selectedNode annotation reset for PVC %q", pvc.Name)
+		}
+
+		if !bound {
+			return false, nil
+		}
+	}
+
+	// All pvs and pvcs that we operated on are bound
+	glog.V(4).Infof("All PVCs for pod %q are bound", podName)
+	return true, nil
 }
 
-func getPVCName(pvc *v1.PersistentVolumeClaim) string {
-	return pvc.Namespace + "/" + pvc.Name
+// check that claimsToBind + claimsToProvision should include all of the Pod's unbound PVCs
```

Comment:
Can this happen that number of (claimsToBind + claimsToProvision) less than a Pod's unbound PVCs?

Reply:
I think a user can delete and recreate a previously bound PVC.  If it got recreated in between predicate and assume, or between assume and bind, then I want to avoid returning binding success if not all PVCs are bound.  But this check isn't race-free.  A user could do the same thing after scheduling and before kubelet comes up.  I guess considering that, we could simplify the logic a little and remove this check.

Reply:
If the pvc deleted , then it need  wait for timeout time (10 minutes ) to return fail? 

```
return wait.Poll(time.Second, b.bindTimeout, func() (bool, error) {
		return b.checkBindings(assumedPod, bindings, claimsToProvision)
	})
```

Reply:
[This check](https://github.com/kubernetes/kubernetes/pull/67556/commits/45d15d4bd2254517ce25e240a3ff9764b6990127#diff-8b07987b1af2d0e15a272411da722f5cR367) returns an error immediately if PVC is deleted.

Similar check for PV deletion too

Reply:
@msau42  But wait.Poll will wait for timeout in this condition?
https://github.com/kubernetes/kubernetes/pull/67556/commits/45d15d4bd2254517ce25e240a3ff9764b6990127#diff-8b07987b1af2d0e15a272411da722f5cR286

Reply:
Nope, the Poll func will return when it got an error

Reply:
@lichuqiang   Thanks, then this make sense.